src/conversion.rs

```diff
@@ -178,6 +179,65 @@ pub trait IntoPy<T>: Sized {
     }
 }
 
+/// Defines a conversion from a Rust type to a Python object, which may fail.
+///
+/// It functions similarly to std's [`TryInto`] trait, but requires a [GIL token](Python)
+/// as an argument.
+pub trait IntoPyObject<'py>: Sized {
+    /// The Python output type
+    type Target;
+    /// The type returned in the event of a conversion error.
+    type Error;
+
+    /// Performs the conversion.
+    fn into_pyobject(self, py: Python<'py>) -> Result<Bound<'py, Self::Target>, Self::Error>;
```

Comment:
I have been thinking hard about APIs like `PyAnyMethods::eq` which currently uses `ToPyObject` as a generic bound. This has the unfortunate effect of forcing an unneeded reference count increase, e.g. if `&Bound<'py, PyAny>` is used as the input argument.

I've been wondering if we should instead have something like

```rust
    /// Performs the conversion.
     fn into_pyobject(self, py: Python<'py>) -> Result<impl Borrow<Bound<'py, Self::Target>>, Self::Error>;
```

This way `impl<'a, 'py> IntoPyObject<'py> for &'a Bound<'py, T>` can return `&'a Bound<'py, T>` as the output type, and we can skip the reference count operation.

This would be a slightly bigger change than what's already proposed here, but might be an awesome performance optimization.

Instead of `Borrow<Bound<'py, T>>` we could also define our own sealed trait which we only implement for `Bound` and `&Bound`, if we think the `Borrow` trait is too opaque.

Reply:
That's a good point. I wonder wether we could use `Cow` instead of an opaque type. Definitely something to experiment with.

Reply:
I am open to `Cow`, however my worry of that is the introduction of a slight runtime cost which may be hard to optimise away and not much better than the original extra refcount op. Will need experimentation, I guess.

Reply:
I've experimented a bit with this:

Using a `Cow` like type like this
```rust
pub enum Foo<'a, 'py, T> {
    Borrowed(Borrowed<'a, 'py, T>),
    Bound(Bound<'py, T>),
}
```
is doable, but it involves quite a bit of additional lifetime shinanigans, because we need that additional lifetime on the trait and in all bounds, sometimes with additional restrictions...

Using an opaque type like you suggested hides this lifetime, but is problematic in places where we call into multiple different `IntoPyObject` implementation. A straight forward example would be
```rust
impl<'py> IntoPyObject<'py> for IpAddr {
    type Target = PyAny;
    type Error = PyErr;

    fn into_pyobject(self, py: Python<'py>) -> Result<impl Foo<'py, Self::Target>, Self::Error> {
        match self {
            IpAddr::V4(ip) => ip.into_pyobject(py),
            IpAddr::V6(ip) => ip.into_pyobject(py),
        }
    }
}
```
which fails to compile because the opaque type in assumed to be different, even tho is turns out to be the same in this case. Since that is pretty common, I don't see a way to make this work.

Reply:
I guess one way is to replace the opaque type with an associated type? (Disclaimer: haven't actually attempted this.)

```rust
trait IntoPyObject<'py> {
    type Target;
    type Output: Borrow<Bound<'py, Self::Target>>;  // Or our own trait instead of `Borrow`
    type Error;

    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error>;
}
```

Then I guess that cases like `IpAddr` above can trivially use the same `Output` type, or otherwise build one up from the sub-implementations. All types would then be statically nameable.

Reply:
Of course! That seems to work quite nicely. Not sure why I haven't seen that, seems quite obvious now 😅. I'll clean up my implementation a bit and then push it here as a separate commit.